neuralpredictors/layers/cores/conv2d.py

```diff
@@ -97,91 +102,124 @@ def __init__(
+        self.laplace_padding = laplace_padding
+        self.hidden_dilation = hidden_dilation
+        self.final_nonlinearity = final_nonlinearity
+        self.elu_xshift, self.elu_yshift = elu_shift
+        self.bias = bias
+        self.momentum = momentum
+        self.pad_input = pad_input
+        self.batch_norm = batch_norm
+        self.batch_norm_scale = batch_norm_scale
+        self.independent_bn_bias = independent_bn_bias
         if stack is None:
-            self.stack = range(self.layers)
+            self.stack = range(self.num_layers)
+        else:
+            self.stack = [*range(self.num_layers)[stack:]] if isinstance(stack, int) else stack
+        self.linear = linear
+
+        if depth_separable:
+            self.ConvLayer = DepthSeparableConv2d
+        elif attention_conv:
+            self.ConvLayer = self.AttentionConvWrapper
         else:
-            self.stack = [*range(self.layers)[stack:]] if isinstance(stack, int) else stack
+            self.ConvLayer = nn.Conv2d
+
+        self.features = nn.Sequential()
+        self.add_first_layer()
+        self.add_subsequent_layers()
+        self.initialize()
 
-        # --- first layer
+    def add_first_layer(self):
         layer = OrderedDict()
         layer["conv"] = nn.Conv2d(
-            input_channels,
-            hidden_channels,
-            input_kern,
-            padding=input_kern // 2 if pad_input else 0,
-            bias=bias and not batch_norm,
+            self.input_channels,
+            self.hidden_channels,
+            self.input_kern,
+            padding=self.input_kern // 2 if self.pad_input else 0,
+            bias=self.bias and not self.batch_norm,
         )
-        if batch_norm:
-            if independent_bn_bias:
-                layer["norm"] = nn.BatchNorm2d(hidden_channels, momentum=momentum)
+        if self.batch_norm:
+            if self.independent_bn_bias:
+                layer["norm"] = nn.BatchNorm2d(self.hidden_channels, momentum=self.momentum)
             else:
-                layer["norm"] = nn.BatchNorm2d(hidden_channels, momentum=momentum, affine=bias and batch_norm_scale)
-                if bias:
-                    if not batch_norm_scale:
-                        layer["bias"] = Bias2DLayer(hidden_channels)
-                elif batch_norm_scale:
-                    layer["scale"] = Scale2DLayer(hidden_channels)
-
-        if layers > 1 or final_nonlinearity:
-            layer["nonlin"] = AdaptiveELU(elu_xshift, elu_yshift)
+                layer["norm"] = nn.BatchNorm2d(
+                    self.hidden_channels, momentum=self.momentum, affine=self.bias and self.batch_norm_scale
+                )
+                if self.bias:
+                    if not self.batch_norm_scale:
+                        layer["bias"] = Bias2DLayer(self.hidden_channels)
+                elif self.batch_norm_scale:
+                    layer["scale"] = Scale2DLayer(self.hidden_channels)
```

Comment:
This whole block is shared between this method and the `add_subsequent_layers` method. Therefore I propose putting it into a separate method. `SE2dCore` also uses this exact block of code if I am not mistaken.